### PR TITLE
fix(indev): swap the order of sending indev events and obj events

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

1. Indev usually needs to do some global processing and needs to have a higher priority than obj events. This is consistent with the `feedback_cb` behavior of LVGL v8.
2. When the distance moved after clicking is greater than the scroll threshold, the click event should not be sent to avoid accidental touches.

```c
static void event_cb(lv_event_t* e)
{
    lv_event_code_t code = lv_event_get_code(e);

    if (code == LV_EVENT_CLICKED) {
        LV_LOG_USER("click");
    } else if (code == LV_EVENT_SHORT_CLICKED) {
        LV_LOG_USER("short click");
    } else if (code == LV_EVENT_LONG_PRESSED) {
        LV_LOG_USER("long press");
    } else if (code == LV_EVENT_LONG_PRESSED_REPEAT) {
        LV_LOG_USER("long press repeat");
    }
}

void obj_test()
{
    lv_obj_add_event_cb(lv_scr_act(), event_cb, LV_EVENT_ALL, NULL);
}
```

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by kisvegabor** [comment]
> When the distance moved after clicking is greater than the scroll threshold, the click event should not be sent to avoid accidental touches

This part is not updated in the PR. Is it intentional? 
Anyway, I think we shouldn't do that. If there was scroll, CLICKED is not sent. But if there was not scroll, it's not problem if the coordinates were moved. 

So I approve those PR as it is :slightly_smiling_face:

**Review by kisvegabor — DISMISSED** [review]

File: src/indev/lv_indev.c [hunk 1]
```diff
@@ -1629,19 +1629,33 @@ static void indev_reset_core(lv_indev_t * indev, lv_obj_t * obj)
...
+        /* Not send click or long pressed event if scroll is detected */
+        if(code == LV_EVENT_SHORT_CLICKED ||
+           code == LV_EVENT_CLICKED ||
+           code == LV_EVENT_LONG_PRESSED ||
+           code == LV_EVENT_LONG_PRESSED_REPEAT) {
+            if(LV_ABS(indev->pointer.scroll_sum.x >= indev->scroll_limit) ||
+               LV_ABS(indev->pointer.scroll_sum.y >= indev->scroll_limit)) {
+                return LV_RESULT_OK;
+            }
+        }
```
Comment (FASTSHIFT) [inline src/indev/lv_indev.c:1677]: @kisvegabor This is a modification of the scroll judgment

**Comment by FASTSHIFT** [comment]
@kisvegabor 
For non-scrollable components, if they are pressed and then dragged for more than a certain distance, can it be considered not a click behavior?

Fix before:
https://github.com/user-attachments/assets/1063cfc5-f21f-40f3-9f8e-4a847fe9976a

Fix after:
https://github.com/user-attachments/assets/c006bfd3-709e-4603-8c6b-032dd4dec695

**Review by liamHowatt — DISMISSED** [review]
Seems like a good change to me.

**Comment by liamHowatt** [comment]
Fyi, I could not play the videos.

**Comment by FASTSHIFT** [comment]
> Fyi, I could not play the videos.

I re-uploaded the video.
You can see that before the fix, dragging after clicking still triggers the click event. After the fix, if the drag exceeds the threshold after clicking, it will be judged as an invalid event and intercepted.

**Comment by FASTSHIFT** [comment]
Can you help confirm this? This may affect all interactions.

File: src/indev/lv_indev.c (hunk 1 above)
Comment (kisvegabor) [reply src/indev/lv_indev.c:1677]: Oh, I should have put my glasses on. :sweat_smile: Thanks!

**Comment by kisvegabor** [comment]
I can't see a clear advantage for sending or not sending click in this case, however this change would be a breaking change as the current app's might rely on this behavior. 

If you need really need to handle this in a simple way, what about adding a `dragged` flag to `lv_indev_t` that you can get in the event and `return` if needed?

**Comment by kisvegabor** [comment]
I've just meant this:
```diff
diff --git a/src/indev/lv_indev.c b/src/indev/lv_indev.c
index 52c904e48..d812d483e 100644
--- a/src/indev/lv_indev.c
+++ b/src/indev/lv_indev.c
@@ -1188,6 +1188,7 @@ static void indev_proc_press(lv_indev_t * indev)
             indev->pointer.gesture_sent   = 0;
             indev->pointer.gesture_sum.x  = 0;
             indev->pointer.gesture_sum.y  = 0;
+            indev->pointer.press_moved    = 0;
             indev->pointer.vect.x         = 0;
             indev->pointer.vect.y         = 0;
 
@@ -1211,12 +1212,15 @@ static void indev_proc_press(lv_indev_t * indev)
     /*Calculate the vector and apply a low pass filter: new value = 0.5 * old_value + 0.5 * new_value*/
     indev->pointer.vect.x = indev->pointer.act_point.x - indev->pointer.last_point.x;
     indev->pointer.vect.y = indev->pointer.act_point.y - indev->pointer.last_point.y;
-
     indev->pointer.scroll_throw_vect.x = (indev->pointer.scroll_throw_vect.x + indev->pointer.vect.x) / 2;
     indev->pointer.scroll_throw_vect.y = (indev->pointer.scroll_throw_vect.y + indev->pointer.vect.y) / 2;
 
     indev->pointer.scroll_throw_vect_ori = indev->pointer.scroll_throw_vect;
 
+    if(LV_ABS(indev->pointer.vect.x) > indev->scroll_limit || LV_ABS(indev->pointer.vect.y) > indev->scroll_limit) {
+       indev->pointer.press_moved = 1;
+    }
+
     if(indev_obj_act) {
         const bool is_enabled = !lv_obj_has_state(indev_obj_act, LV_STATE_DISABLED);
 
diff --git a/src/indev/lv_indev_private.h b/src/indev/lv_indev_private.h
index 04a37e801..571b37b3a 100644
--- a/src/indev/lv_indev_private.h
+++ b/src/indev/lv_indev_private.h
@@ -94,6 +94,7 @@ struct _lv_indev_t {
         lv_dir_t scroll_dir : 4;
         lv_dir_t gesture_dir : 4;
         uint8_t gesture_sent : 1;
+        uint8_t press_moved : 1;
     } pointer;
     struct {
         /*Keypad data*/
```

And add `bool lv_indev_get_press_moved(indev)`. A better name for `press_moved` is very welcome :slightly_smiling_face:

**Comment by FASTSHIFT** [comment]
> I've just meant this:
> 
> ```diff
> diff --git a/src/indev/lv_indev.c b/src/indev/lv_indev.c
> index 52c904e48..d812d483e 100644
> --- a/src/indev/lv_indev.c
> +++ b/src/indev/lv_indev.c
> @@ -1188,6 +1188,7 @@ static void indev_proc_press(lv_indev_t * indev)
>              indev->pointer.gesture_sent   = 0;
>              indev->pointer.gesture_sum.x  = 0;
>              indev->pointer.gesture_sum.y  = 0;
> +            indev->pointer.press_moved    = 0;
>              indev->pointer.vect.x         = 0;
>              indev->pointer.vect.y         = 0;
>  
> @@ -1211,12 +1212,15 @@ static void indev_proc_press(lv_indev_t * indev)
>      /*Calculate the vector and apply a low pass filter: new value = 0.5 * old_value + 0.5 * new_value*/
>      indev->pointer.vect.x = indev->pointer.act_point.x - indev->pointer.last_point.x;
>      indev->pointer.vect.y = indev->pointer.act_point.y - indev->pointer.last_point.y;
> -
>      indev->pointer.scroll_throw_vect.x = (indev->pointer.scroll_throw_vect.x + indev->pointer.vect.x) / 2;
>      indev->pointer.scroll_throw_vect.y = (indev->pointer.scroll_throw_vect.y + indev->pointer.vect.y) / 2;
>  
>      indev->pointer.scroll_throw_vect_ori = indev->pointer.scroll_throw_vect;
>  
> +    if(LV_ABS(indev->pointer.vect.x) > indev->scroll_limit || LV_ABS(indev->pointer.vect.y) > indev->scroll_limit) {
> +       indev->pointer.press_moved = 1;
> +    }
> +
>      if(indev_obj_act) {
>          const bool is_enabled = !lv_obj_has_state(indev_obj_act, LV_STATE_DISABLED);
>  
> diff --git a/src/indev/lv_indev_private.h b/src/indev/lv_indev_private.h
> index 04a37e801..571b37b3a 100644
> --- a/src/indev/lv_indev_private.h
> +++ b/src/indev/lv_indev_private.h
> @@ -94,6 +94,7 @@ struct _lv_indev_t {
>          lv_dir_t scroll_dir : 4;
>          lv_dir_t gesture_dir : 4;
>          uint8_t gesture_sent : 1;
> +        uint8_t press_moved : 1;
>      } pointer;
>      struct {
>          /*Keypad data*/
> ```
> 
> And add `bool lv_indev_get_press_moved(indev)`. A better name for `press_moved` is very welcome 🙂

Thank you very much for your reply. I added `press_moved`. Do you mean to add return in send_event?
```c
static lv_result_t send_event(lv_event_code_t code, void * param)
{
    lv_indev_t * indev = indev_act;

    if(code == LV_EVENT_PRESSED ||
       code == LV_EVENT_SHORT_CLICKED ||
       code == LV_EVENT_CLICKED ||
       code == LV_EVENT_RELEASED ||
       code == LV_EVENT_LONG_PRESSED ||
       code == LV_EVENT_LONG_PRESSED_REPEAT ||
       code == LV_EVENT_ROTARY) {
        lv_indev_send_event(indev, code, indev_obj_act);
        if(indev_reset_check(indev)) return LV_RESULT_INVALID;

        /* Not send click or long pressed event if scroll is detected */
        if(code == LV_EVENT_SHORT_CLICKED ||
           code == LV_EVENT_CLICKED ||
           code == LV_EVENT_LONG_PRESSED ||
           code == LV_EVENT_LONG_PRESSED_REPEAT) {
            if(lv_indev_get_press_moved(indev)) {
                return LV_RESULT_OK;
            }
        }
    }

    lv_obj_send_event(indev_obj_act, code, param);
    if(indev_reset_check(indev)) return LV_RESULT_INVALID;

    return LV_RESULT_OK;
}
```

**Comment by FASTSHIFT** [comment]
Or add interception logic at the application layer:
```c
static void indev_event_cb(lv_event_t* e)
{
    lv_indev_t* indev = (lv_indev_t*)lv_event_get_target(e);
    lv_event_code_t code = lv_event_get_code(e);

    if (code == LV_EVENT_SHORT_CLICKED || code == LV_EVENT_CLICKED || code == LV_EVENT_LONG_PRESSED || code == LV_EVENT_LONG_PRESSED_REPEAT) {
        if (lv_indev_get_press_moved(indev)) {
            return lv_indev_reset(indev, NULL);
        }
    }
}

void obj_test()
{
    lv_obj_add_event_cb(lv_scr_act(), event_cb, LV_EVENT_ALL, NULL);
    lv_indev_add_event_cb(lv_indev_get_next(NULL), indev_event_cb, LV_EVENT_ALL, NULL);
}
```

**Comment by FASTSHIFT** [comment]
How about we add `stop_processing` in indev and modify the `send_event` function like this:

```c
static lv_result_t send_event(lv_event_code_t code, void * param)
{
    lv_indev_t * indev = indev_act;

    if(code == LV_EVENT_PRESSED ||
       code == LV_EVENT_SHORT_CLICKED ||
       code == LV_EVENT_CLICKED ||
       code == LV_EVENT_RELEASED ||
       code == LV_EVENT_LONG_PRESSED ||
       code == LV_EVENT_LONG_PRESSED_REPEAT ||
       code == LV_EVENT_ROTARY) {
        lv_indev_send_event(indev, code, indev_obj_act);
        if(indev_reset_check(indev)) return LV_RESULT_INVALID;
        if(indev_stop_processing_check(indev)) return LV_RESULT_OK;
    }

    lv_obj_send_event(indev_obj_act, code, param);
    if(indev_reset_check(indev)) return LV_RESULT_INVALID;

    return LV_RESULT_OK;
}
```

File: src/indev/lv_indev.h [hunk 2]
```diff
@@ -194,6 +194,13 @@ void * lv_indev_get_user_data(const lv_indev_t * indev);
  */
 void * lv_indev_get_driver_data(const lv_indev_t * indev);
 
+/**
+ * Get whether indev is moved after being clicked
```
Comment (kisvegabor) [inline src/indev/lv_indev.h]:
```suggestion
 * Get whether indev is moved while pressed
```

Comment (FASTSHIFT) [reply src/indev/lv_indev.h]: Done

**Comment by FASTSHIFT** [comment]
> How about we add `stop_processing` in indev and modify the `send_event` function like this:
> 
> ```c
> static lv_result_t send_event(lv_event_code_t code, void * param)
> {
>     lv_indev_t * indev = indev_act;
> 
>     if(code == LV_EVENT_PRESSED ||
>        code == LV_EVENT_SHORT_CLICKED ||
>        code == LV_EVENT_CLICKED ||
>        code == LV_EVENT_RELEASED ||
>        code == LV_EVENT_LONG_PRESSED ||
>        code == LV_EVENT_LONG_PRESSED_REPEAT ||
>        code == LV_EVENT_ROTARY) {
>         lv_indev_send_event(indev, code, indev_obj_act);
>         if(indev_reset_check(indev)) return LV_RESULT_INVALID;
>         if(indev_stop_processing_check(indev)) return LV_RESULT_OK;
>     }
> 
>     lv_obj_send_event(indev_obj_act, code, param);
>     if(indev_reset_check(indev)) return LV_RESULT_INVALID;
> 
>     return LV_RESULT_OK;
> }
> ```
> 
> This will not destroy the indev state machine and can also achieve forward compatibility.

@kisvegabor What do you think of this plan? ;)

**Comment by kisvegabor** [comment]
> What do you think of this plan? ;)

Sorry, I'm back to work. :slightly_smiling_face: 

I like this idea. It's nice and unified.

**Comment by FASTSHIFT** [comment]
> > What do you think of this plan? ;)
> 
> Sorry, I'm back to work. 🙂
> 
> I like this idea. It's nice and unified.

Done

File: src/indev/lv_indev.h [hunk 3]
```diff
@@ -194,13 +194,26 @@ void * lv_indev_get_user_data(const lv_indev_t * indev);
...
+
 /**
  * Reset one or all input devices
  * @param indev pointer to an input device to reset or NULL to reset all of them
  * @param obj pointer to an object which triggers the reset.
  */
 void lv_indev_reset(lv_indev_t * indev, lv_obj_t * obj);
 
+/**
+ * Stop processing input device events
```
Comment (kisvegabor) [inline src/indev/lv_indev.h]:
```suggestion
 * Touch and key related events are sent to the input device first and to the widget after that. If this functions called in an indev event, the event won't be sent to the widget. 
```

**Review by kisvegabor — APPROVED** [review]

**Review by XuNeo — APPROVED** [review]

**Pull request merged**